fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -0,0 +1,701 @@
+                Map<ColumnRefOperator, ScalarOperator> viewExprMap = Utils.getColumnRefMap(
+                        rewriteContext.getMvExpression(), rewriteContext.getMvRefFactory());
+
+                if (!Utils.isAlwaysTrue(equalPredicates)) {
+                    Multimap<ScalarOperator, ColumnRefOperator> normalizedMap =
+                            normalizeAndReverseProjection(viewExprMap, rewriteContext, true);
+                    List<ScalarOperator> conjuncts = Utils.extractConjuncts(equalPredicates);
+                    // swapped by query based view ec
+                    List<ScalarOperator> swappedConjuncts = conjuncts.stream().map(conjunct -> {
+                        ColumnRewriter rewriter = new ColumnRewriter(rewriteContext);
+                        return rewriter.rewriteByViewEc(conjunct);
+                    }).collect(Collectors.toList());
+                    List<ScalarOperator> rewrittens = rewriteQueryScalarOpToTarget(swappedConjuncts, normalizedMap,
+                            rewriteContext.getOutputMapping(), rewriteContext.getQueryColumnSet());
+                    if (rewrittens == null) {
+                        return null;
+                    }
+                    equalPredicates = Utils.compoundAnd(rewrittens);
+                }
+
+                if (!Utils.isAlwaysTrue(otherPredicates)) {
+                    List<ScalarOperator> conjuncts = Utils.extractConjuncts(otherPredicates);
+                    // swapped by query ec
+                    List<ScalarOperator> swappedConjuncts = conjuncts.stream().map(conjunct -> {
+                        ColumnRewriter rewriter = new ColumnRewriter(rewriteContext);
+                        return rewriter.rewriteByQueryEc(conjunct);
+                    }).collect(Collectors.toList());
+                    Multimap<ScalarOperator, ColumnRefOperator> normalizedMap =
+                            normalizeAndReverseProjection(viewExprMap, rewriteContext, false);
+                    List<ScalarOperator> rewrittens = rewriteQueryScalarOpToTarget(swappedConjuncts, normalizedMap,
+                            rewriteContext.getOutputMapping(), rewriteContext.getQueryColumnSet());
+                    if (rewrittens == null) {
+                        return null;
+                    }
+                    otherPredicates = Utils.compoundAnd(rewrittens);
+                }
+            }
+            ScalarOperator compensationPredicate = Utils.canonizePredicate(Utils.compoundAnd(equalPredicates, otherPredicates));
+            if (!Utils.isAlwaysTrue(compensationPredicate)) {
+                // add filter operator
+                LogicalFilterOperator filter = new LogicalFilterOperator(compensationPredicate);
+                rewrittenExpression = OptExpression.create(filter, rewrittenExpression);
+                deriveLogicalProperty(rewrittenExpression);
+            }
```

Comment:
I have question, if the tree before rewite:
```
  Node
     |
 SCAN
```

after rewrite, it's will be?
```
 Node
     |
 Filter
     |
   MV
```

Reply:
yes

Reply:
ok, I see, it's maybe a question later